### ISSUE TITLE
Remove build artifacts from ckanext-dcat_usmetadata repository

### ISSUE THREAD

**Issue by starsinmypockets** [GSA/datagov-ckan-multi#388]
## Acceptance criteria
- [ ] Packaged build files (_main.chunk.HASH.js_ etc) are generated during the build process and are not included as part of the _ckanext-dcat_usmetadata_ repository
- [ ] Built files are correctly loaded via CKAN template

## Tasks
- [ ] Script build file deployment at datagov-deploy
- [ ] Determine naming convention for package files
- [ ] Script template updates as needed

**Comment by adborden** [comment]
Thanks @starsinmypockets, as for the sketch let's keep the "interface" between app and platform (datagov-deploy) clean. That means the compiled assets should be included in the python package, but not in source. Instead, I suggest swapping out the datagov-deploy step for:
- [ ] Add build/package task to ckanext-dcat_usmetadata
- [ ] Publish to pypi from CI on pushes to master
- [ ] Update requirements.txt in inventory-app to pull from pypi instead of git

**Comment by starsinmypockets** [comment]
This all sounds great... I'll prob reach out to @woodt for help on this

**Comment by anuveyatsu** [comment]
@adborden this is great! I have few questions:

1. We have 2 sets of tasks here. Which one should we use? I believe I should refer to the task list from your comment.
2. Do you think we should publish to PyPI on commit to master or on tagged release?
3. Do you believe that removing build artifacts is necessary here? I can't see any issue with it if we're building in the CI.

**Comment by adborden** [comment]
> @adborden this is great! I have few questions:
> 
> 1. We have 2 sets of tasks here. Which one should we use? I believe I should refer to the task list from your comment.

Maybe both sets. I'm not sure what @starsinmypockets had in mind about the build script and handling the names of the artifacts. That might still be necessary. So I would check with him and make sure we combine the tasks into a single list as necessary.

> 2. Do you think we should publish to PyPI on commit to master or on tagged release?

Oh, good point. I prefer tags because then a human can specify a specific version and use semantic versioning.

> 3. Do you believe that removing build artifacts is necessary here? I can't see any issue with it if we're building in the CI.

Not sure I understand the question. I don't want to include any built or compiled files in version control. These are _outputs_ of code, not part of the code. It bloats the repository and makes it more difficult to search files with common tools like `git grep`. These should be git ignored, but included in the MANIFEST so they are included in the python package that gets published. Then GSA/inventory-app can pull from pypi in the requirements.txt instead of the git repo. Does that answer your question?

**Comment by anuveyatsu** [comment]
@adborden to make sure that we're on the same page, I created a sequence diagram of the process so we can visually (and easily) understand it. You might need to have "mermaid + github" chrome extension to see the diagram below. I'm also pasting the screenshot of the diagram just in case.

I have also created a new set of tasks / acceptance criteria to consolidate yours and Paul's. Hope this makes sense for you.

## Diagram of the desired steps

```mermaid
sequenceDiagram
    Developer->>Git: Push tagged commit to master branch
    Git-->>CI/CD: Trigger deployment
    CI/CD-->>CI/CD: Run tests
    CI/CD-->>CI/CD: Build assets (JS bundles)
    CI/CD-->>PyPI: Publish the package
    Inventory-->>PyPI: Install
```

The screenshot of above sequence diagram:

![Screen Shot 2020-10-15 at 20.23.20.png](https://images.zenhubusercontent.com/583c185ab9dc3c622022968d/635dd713-1fe5-45ec-943d-4255c06f562b)

## Acceptance criteria

* [ ] Builds aren't committed into the git repo
* [ ] I can install the package from the PyPI using semantic versioning

## Tasks

* [ ] Create CI job that runs on tagged commits:
    * [ ] Depends on test jobs and runs only if test jobs are completed
    * [ ] Builds the assets
    * [ ] Publishes to PyPI with the assets
* [ ] Update README explaining the deployment/publishing process
* [ ] Remove bundles from the git repo
    * [ ] Update the README explaining this and what should be done to get the extension working locally if pulling from git repo
    * [ ] Update `.gitingore` file so bundles are ignored

**Comment by adborden** [comment]
Yes, the only thing I would change is the JS build needs to happen before the tests, right?

```
sequenceDiagram
    Developer->>Git: Push tagged commit to master branch
    Git-->>CI/CD: Trigger deployment
    CI/CD-->>CI/CD: Build assets (JS bundles)
    CI/CD-->>CI/CD: Build python package
    CI/CD-->>CI/CD: Run tests
    CI/CD-->>PyPI: Publish the package
    Inventory-->>PyPI: Install
```

[![Sequence diagram](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gICAgRGV2ZWxvcGVyLT4-R2l0OiBQdXNoIHRhZ2dlZCBjb21taXQgdG8gbWFzdGVyIGJyYW5jaFxuICAgIEdpdC0tPj5DSS9DRDogVHJpZ2dlciBkZXBsb3ltZW50XG4gICAgQ0kvQ0QtLT4-Q0kvQ0Q6IEJ1aWxkIGFzc2V0cyAoSlMgYnVuZGxlcylcbiAgICBDSS9DRC0tPj5DSS9DRDogQnVpbGQgcHl0aG9uIHBhY2thZ2VcbiAgICBDSS9DRC0tPj5DSS9DRDogUnVuIHRlc3RzXG4gICAgQ0kvQ0QtLT4-UHlQSTogUHVibGlzaCB0aGUgcGFja2FnZVxuICAgIEludmVudG9yeS0tPj5QeVBJOiBJbnN0YWxsIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQiLCJ0aGVtZVZhcmlhYmxlcyI6eyJiYWNrZ3JvdW5kIjoid2hpdGUiLCJwcmltYXJ5Q29sb3IiOiIjRUNFQ0ZGIiwic2Vjb25kYXJ5Q29sb3IiOiIjZmZmZmRlIiwidGVydGlhcnlDb2xvciI6ImhzbCg4MCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwicHJpbWFyeUJvcmRlckNvbG9yIjoiaHNsKDI0MCwgNjAlLCA4Ni4yNzQ1MDk4MDM5JSkiLCJzZWNvbmRhcnlCb3JkZXJDb2xvciI6ImhzbCg2MCwgNjAlLCA4My41Mjk0MTE3NjQ3JSkiLCJ0ZXJ0aWFyeUJvcmRlckNvbG9yIjoiaHNsKDgwLCA2MCUsIDg2LjI3NDUwOTgwMzklKSIsInByaW1hcnlUZXh0Q29sb3IiOiIjMTMxMzAwIiwic2Vjb25kYXJ5VGV4dENvbG9yIjoiIzAwMDAyMSIsInRlcnRpYXJ5VGV4dENvbG9yIjoicmdiKDkuNTAwMDAwMDAwMSwgOS41MDAwMDAwMDAxLCA5LjUwMDAwMDAwMDEpIiwibGluZUNvbG9yIjoiIzMzMzMzMyIsInRleHRDb2xvciI6IiMzMzMiLCJtYWluQmtnIjoiI0VDRUNGRiIsInNlY29uZEJrZyI6IiNmZmZmZGUiLCJib3JkZXIxIjoiIzkzNzBEQiIsImJvcmRlcjIiOiIjYWFhYTMzIiwiYXJyb3doZWFkQ29sb3IiOiIjMzMzMzMzIiwiZm9udEZhbWlseSI6IlwidHJlYnVjaGV0IG1zXCIsIHZlcmRhbmEsIGFyaWFsIiwiZm9udFNpemUiOiIxNnB4IiwibGFiZWxCYWNrZ3JvdW5kIjoiI2U4ZThlOCIsIm5vZGVCa2ciOiIjRUNFQ0ZGIiwibm9kZUJvcmRlciI6IiM5MzcwREIiLCJjbHVzdGVyQmtnIjoiI2ZmZmZkZSIsImNsdXN0ZXJCb3JkZXIiOiIjYWFhYTMzIiwiZGVmYXVsdExpbmtDb2xvciI6IiMzMzMzMzMiLCJ0aXRsZUNvbG9yIjoiIzMzMyIsImVkZ2VMYWJlbEJhY2tncm91bmQiOiIjZThlOGU4IiwiYWN0b3JCb3JkZXIiOiJoc2woMjU5LjYyNjE2ODIyNDMsIDU5Ljc3NjUzNjMxMjglLCA4Ny45MDE5NjA3ODQzJSkiLCJhY3RvckJrZyI6IiNFQ0VDRkYiLCJhY3RvclRleHRDb2xvciI6ImJsYWNrIiwiYWN0b3JMaW5lQ29sb3IiOiJncmV5Iiwic2lnbmFsQ29sb3IiOiIjMzMzIiwic2lnbmFsVGV4dENvbG9yIjoiIzMzMyIsImxhYmVsQm94QmtnQ29sb3IiOiIjRUNFQ0ZGIiwibGFiZWxCb3hCb3JkZXJDb2xvciI6ImhzbCgyNTkuNjI2MTY4MjI0MywgNTkuNzc2NTM2MzEyOCUsIDg3LjkwMTk2MDc4NDMlKSIsImxhYmVsVGV4dENvbG9yIjoiYmxhY2siLCJsb29wVGV4dENvbG9yIjoiYmxhY2siLCJub3RlQm9yZGVyQ29sb3IiOiIjYWFhYTMzIiwibm90ZUJrZ0NvbG9yIjoiI2ZmZjVhZCIsIm5vdGVUZXh0Q29sb3IiOiJibGFjayIsImFjdGl2YXRpb25Cb3JkZXJDb2xvciI6IiM2NjYiLCJhY3RpdmF0aW9uQmtnQ29sb3IiOiIjZjRmNGY0Iiwic2VxdWVuY2VOdW1iZXJDb2xvciI6IndoaXRlIiwic2VjdGlvbkJrZ0NvbG9yIjoicmdiYSgxMDIsIDEwMiwgMjU1LCAwLjQ5KSIsImFsdFNlY3Rpb25Ca2dDb2xvciI6IndoaXRlIiwic2VjdGlvbkJrZ0NvbG9yMiI6IiNmZmY0MDAiLCJ0YXNrQm9yZGVyQ29sb3IiOiIjNTM0ZmJjIiwidGFza0JrZ0NvbG9yIjoiIzhhOTBkZCIsInRhc2tUZXh0TGlnaHRDb2xvciI6IndoaXRlIiwidGFza1RleHRDb2xvciI6IndoaXRlIiwidGFza1RleHREYXJrQ29sb3IiOiJibGFjayIsInRhc2tUZXh0T3V0c2lkZUNvbG9yIjoiYmxhY2siLCJ0YXNrVGV4dENsaWNrYWJsZUNvbG9yIjoiIzAwMzE2MyIsImFjdGl2ZVRhc2tCb3JkZXJDb2xvciI6IiM1MzRmYmMiLCJhY3RpdmVUYXNrQmtnQ29sb3IiOiIjYmZjN2ZmIiwiZ3JpZENvbG9yIjoibGlnaHRncmV5IiwiZG9uZVRhc2tCa2dDb2xvciI6ImxpZ2h0Z3JleSIsImRvbmVUYXNrQm9yZGVyQ29sb3IiOiJncmV5IiwiY3JpdEJvcmRlckNvbG9yIjoiI2ZmODg4OCIsImNyaXRCa2dDb2xvciI6InJlZCIsInRvZGF5TGluZUNvbG9yIjoicmVkIiwibGFiZWxDb2xvciI6ImJsYWNrIiwiZXJyb3JCa2dDb2xvciI6IiM1NTIyMjIiLCJlcnJvclRleHRDb2xvciI6IiM1NTIyMjIiLCJjbGFzc1RleHQiOiIjMTMxMzAwIiwiZmlsbFR5cGUwIjoiI0VDRUNGRiIsImZpbGxUeXBlMSI6IiNmZmZmZGUiLCJmaWxsVHlwZTIiOiJoc2woMzA0LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTMiOiJoc2woMTI0LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkiLCJmaWxsVHlwZTQiOiJoc2woMTc2LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTUiOiJoc2woLTQsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSIsImZpbGxUeXBlNiI6ImhzbCg4LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTciOiJoc2woMTg4LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkifX0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gICAgRGV2ZWxvcGVyLT4-R2l0OiBQdXNoIHRhZ2dlZCBjb21taXQgdG8gbWFzdGVyIGJyYW5jaFxuICAgIEdpdC0tPj5DSS9DRDogVHJpZ2dlciBkZXBsb3ltZW50XG4gICAgQ0kvQ0QtLT4-Q0kvQ0Q6IEJ1aWxkIGFzc2V0cyAoSlMgYnVuZGxlcylcbiAgICBDSS9DRC0tPj5DSS9DRDogQnVpbGQgcHl0aG9uIHBhY2thZ2VcbiAgICBDSS9DRC0tPj5DSS9DRDogUnVuIHRlc3RzXG4gICAgQ0kvQ0QtLT4-UHlQSTogUHVibGlzaCB0aGUgcGFja2FnZVxuICAgIEludmVudG9yeS0tPj5QeVBJOiBJbnN0YWxsIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQiLCJ0aGVtZVZhcmlhYmxlcyI6eyJiYWNrZ3JvdW5kIjoid2hpdGUiLCJwcmltYXJ5Q29sb3IiOiIjRUNFQ0ZGIiwic2Vjb25kYXJ5Q29sb3IiOiIjZmZmZmRlIiwidGVydGlhcnlDb2xvciI6ImhzbCg4MCwgMTAwJSwgOTYuMjc0NTA5ODAzOSUpIiwicHJpbWFyeUJvcmRlckNvbG9yIjoiaHNsKDI0MCwgNjAlLCA4Ni4yNzQ1MDk4MDM5JSkiLCJzZWNvbmRhcnlCb3JkZXJDb2xvciI6ImhzbCg2MCwgNjAlLCA4My41Mjk0MTE3NjQ3JSkiLCJ0ZXJ0aWFyeUJvcmRlckNvbG9yIjoiaHNsKDgwLCA2MCUsIDg2LjI3NDUwOTgwMzklKSIsInByaW1hcnlUZXh0Q29sb3IiOiIjMTMxMzAwIiwic2Vjb25kYXJ5VGV4dENvbG9yIjoiIzAwMDAyMSIsInRlcnRpYXJ5VGV4dENvbG9yIjoicmdiKDkuNTAwMDAwMDAwMSwgOS41MDAwMDAwMDAxLCA5LjUwMDAwMDAwMDEpIiwibGluZUNvbG9yIjoiIzMzMzMzMyIsInRleHRDb2xvciI6IiMzMzMiLCJtYWluQmtnIjoiI0VDRUNGRiIsInNlY29uZEJrZyI6IiNmZmZmZGUiLCJib3JkZXIxIjoiIzkzNzBEQiIsImJvcmRlcjIiOiIjYWFhYTMzIiwiYXJyb3doZWFkQ29sb3IiOiIjMzMzMzMzIiwiZm9udEZhbWlseSI6IlwidHJlYnVjaGV0IG1zXCIsIHZlcmRhbmEsIGFyaWFsIiwiZm9udFNpemUiOiIxNnB4IiwibGFiZWxCYWNrZ3JvdW5kIjoiI2U4ZThlOCIsIm5vZGVCa2ciOiIjRUNFQ0ZGIiwibm9kZUJvcmRlciI6IiM5MzcwREIiLCJjbHVzdGVyQmtnIjoiI2ZmZmZkZSIsImNsdXN0ZXJCb3JkZXIiOiIjYWFhYTMzIiwiZGVmYXVsdExpbmtDb2xvciI6IiMzMzMzMzMiLCJ0aXRsZUNvbG9yIjoiIzMzMyIsImVkZ2VMYWJlbEJhY2tncm91bmQiOiIjZThlOGU4IiwiYWN0b3JCb3JkZXIiOiJoc2woMjU5LjYyNjE2ODIyNDMsIDU5Ljc3NjUzNjMxMjglLCA4Ny45MDE5NjA3ODQzJSkiLCJhY3RvckJrZyI6IiNFQ0VDRkYiLCJhY3RvclRleHRDb2xvciI6ImJsYWNrIiwiYWN0b3JMaW5lQ29sb3IiOiJncmV5Iiwic2lnbmFsQ29sb3IiOiIjMzMzIiwic2lnbmFsVGV4dENvbG9yIjoiIzMzMyIsImxhYmVsQm94QmtnQ29sb3IiOiIjRUNFQ0ZGIiwibGFiZWxCb3hCb3JkZXJDb2xvciI6ImhzbCgyNTkuNjI2MTY4MjI0MywgNTkuNzc2NTM2MzEyOCUsIDg3LjkwMTk2MDc4NDMlKSIsImxhYmVsVGV4dENvbG9yIjoiYmxhY2siLCJsb29wVGV4dENvbG9yIjoiYmxhY2siLCJub3RlQm9yZGVyQ29sb3IiOiIjYWFhYTMzIiwibm90ZUJrZ0NvbG9yIjoiI2ZmZjVhZCIsIm5vdGVUZXh0Q29sb3IiOiJibGFjayIsImFjdGl2YXRpb25Cb3JkZXJDb2xvciI6IiM2NjYiLCJhY3RpdmF0aW9uQmtnQ29sb3IiOiIjZjRmNGY0Iiwic2VxdWVuY2VOdW1iZXJDb2xvciI6IndoaXRlIiwic2VjdGlvbkJrZ0NvbG9yIjoicmdiYSgxMDIsIDEwMiwgMjU1LCAwLjQ5KSIsImFsdFNlY3Rpb25Ca2dDb2xvciI6IndoaXRlIiwic2VjdGlvbkJrZ0NvbG9yMiI6IiNmZmY0MDAiLCJ0YXNrQm9yZGVyQ29sb3IiOiIjNTM0ZmJjIiwidGFza0JrZ0NvbG9yIjoiIzhhOTBkZCIsInRhc2tUZXh0TGlnaHRDb2xvciI6IndoaXRlIiwidGFza1RleHRDb2xvciI6IndoaXRlIiwidGFza1RleHREYXJrQ29sb3IiOiJibGFjayIsInRhc2tUZXh0T3V0c2lkZUNvbG9yIjoiYmxhY2siLCJ0YXNrVGV4dENsaWNrYWJsZUNvbG9yIjoiIzAwMzE2MyIsImFjdGl2ZVRhc2tCb3JkZXJDb2xvciI6IiM1MzRmYmMiLCJhY3RpdmVUYXNrQmtnQ29sb3IiOiIjYmZjN2ZmIiwiZ3JpZENvbG9yIjoibGlnaHRncmV5IiwiZG9uZVRhc2tCa2dDb2xvciI6ImxpZ2h0Z3JleSIsImRvbmVUYXNrQm9yZGVyQ29sb3IiOiJncmV5IiwiY3JpdEJvcmRlckNvbG9yIjoiI2ZmODg4OCIsImNyaXRCa2dDb2xvciI6InJlZCIsInRvZGF5TGluZUNvbG9yIjoicmVkIiwibGFiZWxDb2xvciI6ImJsYWNrIiwiZXJyb3JCa2dDb2xvciI6IiM1NTIyMjIiLCJlcnJvclRleHRDb2xvciI6IiM1NTIyMjIiLCJjbGFzc1RleHQiOiIjMTMxMzAwIiwiZmlsbFR5cGUwIjoiI0VDRUNGRiIsImZpbGxUeXBlMSI6IiNmZmZmZGUiLCJmaWxsVHlwZTIiOiJoc2woMzA0LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTMiOiJoc2woMTI0LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkiLCJmaWxsVHlwZTQiOiJoc2woMTc2LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTUiOiJoc2woLTQsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSIsImZpbGxUeXBlNiI6ImhzbCg4LCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJmaWxsVHlwZTciOiJoc2woMTg4LCAxMDAlLCA5My41Mjk0MTE3NjQ3JSkifX0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)

**Comment by anuveyatsu** [comment]
@adborden yes, builds should happen before the tests, otherwise, the tests would fail :+1:

**Comment by anuveyatsu** [comment]
@adborden I've got couple of questions:

## Author

I've noticed that in `setup.py` we have Paul Walker's details as an author. Is it expected or should we update it, eg, we can use a US Department/Agency name and email address there:

```
# Author details
author='''Paul Walker''',
author_email='''paul.walker@datopian.com''',
```

Please, let me know wdyt.

## Owner

At the moment, I'm the only maintainer of the project on PyPI, therefore, I'm the owner. But I'd like to add more collaborators for the project and transfer ownership. Could you give me the details (username on PyPI) of the users?

**Comment by adborden** [comment]
Yeah, author should reflect the data.gov team.

author = Data.gov
author_email = datagovhelp at gsa dot gov

For the project ownership, transfer it to me for now, https://pypi.org/user/adborden/.

**Comment by anuveyatsu** [comment]
@adborden author is now updated.

I've just sent an invite in PyPI.

**Comment by adborden** [comment]
@anuveyatsu let me know when you get the build fixed on sandbox and then I will call this done.

I also made a PR to fix an issue with the gitignore ^ but also to use git tags instead of GH releases. We don't use GH releases for any of our projects at the moment.

**Comment by adborden** [comment]
@anuveyatsu there's still some USWDS assets under src/{img,fonts,css}. These should be grabbed from the package under node_modules.

**Comment by adborden** [comment]
Nevermind, using the uswds npm package is called out in this story https://github.com/gsa/datagov-ckan-multi/issues/384

**Comment by anuveyatsu** [comment]
@adborden I'm now about to merge your PR here https://github.com/GSA/ckanext-dcat_usmetadata/pull/118

**Comment by adborden** [comment]
Looks good.
